### ISSUE TITLE
Thanos not reading newly uploaded blocks

### ISSUE THREAD

**Issue by lpashos** [thanos-io/thanos#4933]
**Thanos, Prometheus and Golang version used**:
Thanos: 0.23.1
Prometheus: 2.30.3

**Object Storage Provider**:
Swift

**What happened**:
Thanos sidecar is uploading blocks to object storage, but I cannot query any new blocks uploaded after this past September. I can see newly uploaded blocks in our swift interface directly but thanos store does not seem to be indexing them or allow me to query them. I cannot see them with thanos tools either.

**What you expected to happen**:

Newly uploaded blocks to be indexed and be queryable, or there to be at least some sort of error in the logs pointing to an issue.

**How to reproduce it (as minimally and precisely as possible)**:

**Full logs to relevant components**:

*Last few days of thanos-sidecar logs*

```level=info ts=2021-12-06T01:00:28.291330314Z caller=shipper.go:337 msg="upload new block" id=01FP6M2CNMDEA12QYM4XAAAZBR
level=info ts=2021-12-06T03:00:28.110203518Z caller=shipper.go:337 msg="upload new block" id=01FP6TY3XN05B1CVJXR6M7421D
level=info ts=2021-12-06T05:00:28.109532504Z caller=shipper.go:337 msg="upload new block" id=01FP71SV5NVXDEFD1T9J0WQEF7
level=info ts=2021-12-06T07:00:58.109686297Z caller=shipper.go:337 msg="upload new block" id=01FP78NJE3BH8E466YHH913X3C
level=info ts=2021-12-06T09:00:58.108385896Z caller=shipper.go:337 msg="upload new block" id=01FP7FH9NNR3TF8BCXFXCF94YV
level=info ts=2021-12-06T11:00:58.117311758Z caller=shipper.go:337 msg="upload new block" id=01FP7PD0XNMWPW2K7XSVHMN4N4
level=info ts=2021-12-06T13:00:28.310070192Z caller=shipper.go:337 msg="upload new block" id=01FP7X8R5NP3AEBBBV53GCFNN0
level=info ts=2021-12-06T15:00:28.108354385Z caller=shipper.go:337 msg="upload new block" id=01FP844FDNN0EHSV37JYC77FE6
level=info ts=2021-12-06T17:00:28.113764457Z caller=shipper.go:337 msg="upload new block" id=01FP8B06NS55PR45G4S9X4ACA9
level=info ts=2021-12-06T19:00:58.10643592Z caller=shipper.go:337 msg="upload new block" id=01FP8HVXXNNWRHKGD03KVN4RWR
level=info ts=2021-12-06T21:00:28.109912427Z caller=shipper.go:337 msg="upload new block" id=01FP8RQN5NXM14X7KV8BEAKGF3
level=info ts=2021-12-06T23:00:58.117373136Z caller=shipper.go:337 msg="upload new block" id=01FP8ZKCDN60YGW0SM4V8N65PJ
level=info ts=2021-12-07T01:00:58.1994395Z caller=shipper.go:337 msg="upload new block" id=01FP96F3NN42G3Z8HQWSAA56ZB
level=info ts=2021-12-07T03:00:28.10884307Z caller=shipper.go:337 msg="upload new block" id=01FP9DATXMWAPWWJSJRPMQ2Y33
level=info ts=2021-12-07T05:00:58.108401396Z caller=shipper.go:337 msg="upload new block" id=01FP9M6J5NPX5NFTSE83EQHEQ5
level=info ts=2021-12-07T07:00:28.109938466Z caller=shipper.go:337 msg="upload new block" id=01FP9V29DRQJ69DFBGBK19X8TM
level=info ts=2021-12-07T09:00:28.112379167Z caller=shipper.go:337 msg="upload new block" id=01FPA1Y0NP7PGW74HRT5S2A2WT
level=info ts=2021-12-07T11:00:28.110892707Z caller=shipper.go:337 msg="upload new block" id=01FPA8SQXMW5S5A1GGWZWR9J7G
level=info ts=2021-12-07T13:00:58.212321497Z caller=shipper.go:337 msg="upload new block" id=01FPAFNF5PNMHY2FJSB7ZK52DR
level=info ts=2021-12-07T15:00:58.126902387Z caller=shipper.go:337 msg="upload new block" id=01FPAPH6DNE4RXJ40BN4MN11BE
level=info ts=2021-12-07T17:00:58.107641604Z caller=shipper.go:337 msg="upload new block" id=01FPAXCXNM2EKDP4DPGVFWV9NZ
level=info ts=2021-12-07T19:00:58.107148304Z caller=shipper.go:337 msg="upload new block" id=01FPB48MY1WASDFWMEZF8KJ5QC
level=info ts=2021-12-07T21:00:28.11156975Z caller=shipper.go:337 msg="upload new block" id=01FPBB4C5NZGXFQ19J01NFHZQP
level=info ts=2021-12-07T23:00:58.120359257Z caller=shipper.go:337 msg="upload new block" id=01FPBJ03DPPRAH11RXTQDYMYBD
level=info ts=2021-12-08T01:00:58.215836714Z caller=shipper.go:337 msg="upload new block" id=01FPBRVTNPPHFXGSXTHS4TA9XD
level=info ts=2021-12-08T03:00:28.108025135Z caller=shipper.go:337 msg="upload new block" id=01FPBZQHXNX2AD0ZD4RFQRDERG
level=info ts=2021-12-08T05:00:28.108162267Z caller=shipper.go:337 msg="upload new block" id=01FPC6K95N6J8CE369P7KP7MP2
level=info ts=2021-12-08T07:00:58.10767824Z caller=shipper.go:337 msg="upload new block" id=01FPCDF0DS17DPSTJWQFTED5HK
level=info ts=2021-12-08T09:00:28.107857791Z caller=shipper.go:337 msg="upload new block" id=01FPCMAQNNXFW1187704D7HJ32
level=info ts=2021-12-08T11:00:28.118519438Z caller=shipper.go:337 msg="upload new block" id=01FPCV6EXNPH7YD8HPTSNMCEPZ
level=info ts=2021-12-08T13:00:58.38082914Z caller=shipper.go:337 msg="upload new block" id=01FPD2265NYJKNZS1B6HK0ZY2N
level=info ts=2021-12-08T15:00:58.108467506Z caller=shipper.go:337 msg="upload new block" id=01FPD8XXDNG0GD41SED6883D1V
```

*thanos-store logs repeat block metadata sync with same number of blocks*

```
level=info ts=2021-12-03T17:24:26.2140222Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=541.723393ms duration_ms=541 cached=206 returned=206 partial=0
level=info ts=2021-12-06T18:12:26.210296479Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=538.04282ms duration_ms=538 cached=206 returned=206 partial=0
level=info ts=2021-12-07T19:45:26.235772458Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=560.988617ms duration_ms=560 cached=206 returned=206 partial=0
level=info ts=2021-12-08T15:45:26.12836438Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=456.659276ms duration_ms=456 cached=206 returned=206 partial=0
```

*"thanos tools bucket inspect" shows no new blocks since September:*

```root@thanos-a-1:~/thanos-0.23.1.linux-amd64# ./thanos tools bucket inspect --objstore.config-file=/etc/prometheus/thanos_objstore.yml | grep 09-2021
level=info ts=2021-12-08T14:51:11.951671219Z caller=factory.go:49 msg="loading bucket configuration"
level=info ts=2021-12-08T14:51:12.566868659Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=494.223249ms duration_ms=494 cached=206 returned=206 partial=0
| 01FEJHHR8JTWEX9XAFVJ6FMH6K | 19-08-2021 00:00:00 | 02-09-2021 00:00:00 | 335h59m59.979s | -295h59m59.979s | 3,427,072 | 96,755,734,160 | 806,988,550 | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 0s         | compactor |
| 01FEJPM4JV0ZEY3SPPXH8Y0ABW | 19-08-2021 00:00:00 | 02-09-2021 00:00:00 | 335h59m59.979s | -95h59m59.979s  | 3,427,054 | 11,947,448,239 | 86,376,862  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 5m0s       | compactor |
| 01FEJX79VYMM5C5MG0AXZ81VG3 | 19-08-2021 00:00:00 | 02-09-2021 00:00:00 | 335h59m59.979s | -               | 3,427,054 | 997,332,376    | 12,297,715  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 1h0m0s     | compactor |
| 01FEJNZQDVB592GEV253NEVD0W | 19-08-2021 00:00:02 | 02-09-2021 00:00:00 | 335h59m57.86s  | -295h59m57.86s  | 7,851     | 165,276,144    | 1,412,738   | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 0s         | compactor |
| 01FEJWXDQY7DQ63XH136YCKTPN | 19-08-2021 00:00:02 | 02-09-2021 00:00:00 | 335h59m57.86s  | -95h59m57.86s   | 7,851     | 16,578,072     | 122,515     | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 5m0s       | compactor |
| 01FEJZ74B5HE0C6ZXM3KNXNAPN | 19-08-2021 00:00:02 | 02-09-2021 00:00:00 | 335h59m57.86s  | -               | 7,851     | 1,420,543      | 20,132      | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 1h0m0s     | compactor |
| 01FFPJZ84RRF2GCDD5CMNJ64FP | 02-09-2021 00:00:00 | 16-09-2021 00:00:00 | 335h59m59.979s | -295h59m59.979s | 3,400,870 | 97,098,335,993 | 809,920,519 | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 0s         | compactor |
| 01FFPR0VMMDR6TDMTVSZYBAMZ5 | 02-09-2021 00:00:00 | 16-09-2021 00:00:00 | 335h59m59.979s | -95h59m59.979s  | 3,400,852 | 12,002,665,414 | 86,740,615  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 5m0s       | compactor |
| 01FFPZ6C55PK4YTQAE4ZCY5R3W | 02-09-2021 00:00:00 | 16-09-2021 00:00:00 | 335h59m59.979s | -               | 3,400,852 | 1,001,912,047  | 12,322,824  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 1h0m0s     | compactor |
| 01FFPQBTDGV4VT5NMQSHKA5PM5 | 02-09-2021 00:00:02 | 16-09-2021 00:00:00 | 335h59m57.86s  | -295h59m57.86s  | 7,020     | 165,228,294    | 1,409,521   | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 0s         | compactor |
| 01FFPYWMW0WXFNJ1HPR5CH0QKP | 02-09-2021 00:00:02 | 16-09-2021 00:00:00 | 335h59m57.86s  | -95h59m57.86s   | 7,020     | 16,582,692     | 121,682     | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 5m0s       | compactor |
| 01FFQ1A69PAM9QXK3ZAR3F5GHD | 02-09-2021 00:00:02 | 16-09-2021 00:00:00 | 335h59m57.86s  | -               | 7,020     | 1,417,131      | 19,269      | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 1h0m0s     | compactor |
| 01FGTMZ87NGDD1Y58K2VRQAVC4 | 16-09-2021 00:00:00 | 30-09-2021 00:00:00 | 335h59m59.979s | -295h59m59.979s | 3,412,627 | 97,466,563,770 | 812,981,973 | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 0s         | compactor |
| 01FGTV1TDPMMQ483C9SXBGTVSH | 16-09-2021 00:00:00 | 30-09-2021 00:00:00 | 335h59m59.979s | -95h59m59.979s  | 3,412,609 | 12,052,502,447 | 87,094,627  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 5m0s       | compactor |
| 01FGV2N00S9JQ5WMB61JNBVAT1 | 16-09-2021 00:00:00 | 30-09-2021 00:00:00 | 335h59m59.979s | -               | 3,412,609 | 1,006,234,608  | 12,368,039  | 4          | false       | environment=Production,prometheus_server=thanos-a-1.phl,service=Prometheus | 1h0m0s     | compactor |
| 01FGTT9N9J2G68H8MTNR0N7S9C | 16-09-2021 00:00:02 | 30-09-2021 00:00:00 | 335h59m57.86s  | -295h59m57.86s  | 26,302    | 542,236,317    | 4,558,766   | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 0s         | compactor |
| 01FGV2A8C2VVTED6RFDB2R6979 | 16-09-2021 00:00:02 | 30-09-2021 00:00:00 | 335h59m57.86s  | -95h59m57.86s   | 26,302    | 54,289,362     | 395,947     | 4          | false       | prometheus_server=thanos-a-1.phl-rule                                      | 5m0s       | compactor |
level=info ts=2021-12-08T14:51:12.585586767Z caller=main.go:160 msg=exiting

root@thanos-a-1:~/thanos-0.23.1.linux-amd64# ./thanos tools bucket inspect --objstore.config-file=/etc/prometheus/thanos_objstore.yml | grep 10-2021
level=info ts=2021-12-08T14:52:45.600288241Z caller=factory.go:49 msg="loading bucket configuration"
level=info ts=2021-12-08T14:52:46.361983507Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=463.397051ms duration_ms=463 cached=206 returned=206 partial=0
level=info ts=2021-12-08T14:52:46.384084705Z caller=main.go:160 msg=exiting

root@thanos-a-1:~/thanos-0.23.1.linux-amd64# ./thanos tools bucket inspect --objstore.config-file=/etc/prometheus/thanos_objstore.yml | grep 11-2021
level=info ts=2021-12-08T14:53:20.663422996Z caller=factory.go:49 msg="loading bucket configuration"
level=info ts=2021-12-08T14:53:21.279779032Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=494.714997ms duration_ms=494 cached=206 returned=206 partial=0
level=info ts=2021-12-08T14:53:21.301678007Z caller=main.go:160 msg=exiting

root@thanos-a-1:~/thanos-0.23.1.linux-amd64# ./thanos tools bucket inspect --objstore.config-file=/etc/prometheus/thanos_objstore.yml | grep 12-2021
level=info ts=2021-12-08T14:53:51.612662217Z caller=factory.go:49 msg="loading bucket configuration"
level=info ts=2021-12-08T14:53:52.347299755Z caller=fetcher.go:476 component=block.BaseFetcher msg="successfully synchronized block metadata" duration=492.484166ms duration_ms=492 cached=206 returned=206 partial=0
level=info ts=2021-12-08T14:53:52.36651175Z caller=main.go:160 msg=exiting
```

*Newly uploaded blocks in my swift interface*
![image](https://user-images.githubusercontent.com/39998504/145236363-f54db67c-56e7-4d15-b184-7033fb2c9017.png)
![image](https://user-images.githubusercontent.com/39998504/145236497-503e6a32-9238-4d4c-a58f-5e3125cf9d9a.png)
![image](https://user-images.githubusercontent.com/39998504/145236675-a67f0260-fa6a-4ebf-ae76-a922fa01116b.png)

**Anything else we need to know**:
Prometheus and all thanos services run on one host
Command strings:
*Prometheus*
```
/usr/bin/docker run --rm --name prometheus --network host --volume /etc/prometheus/:/etc/prometheus/ --volume thanos-data:/opt/bitnami/prometheus/data bitnami/prometheus:2.30.3 --config.file=/etc/prometheus/prometheus.yml --storage.tsdb.max-block-duration=2h --storage.tsdb.min-block-duration=2h --web.enable-lifecycle --web.enable-admin-api --web.external-url=http://thanos-a-1.phl/prometheus --web.route-prefix=/
```
*Thanos Sidecar*
```
/usr/bin/docker run --rm --name thanos-sidecar --network host --volume /etc/prometheus/thanos_objstore.yml:/thanos_objstore.yml --volume thanos-data:/data bitnami/thanos:0.23.1 sidecar --tsdb.path=/data --prometheus.url=http://localhost/prometheus --grpc-address=0.0.0.0:19100 --http-address=0.0.0.0:19200 --objstore.config-file=/thanos_objstore.yml
```
*Thanos Rule*
```
/usr/bin/docker run --rm --name thanos-rule --network host --volume /etc/prometheus/:/etc/prometheus/ --volume thanos-data:/data bitnami/thanos:0.23.1 rule --data-dir=/data/thanos/rule --grpc-address=0.0.0.0:19104 --http-address=0.0.0.0:19204 --objstore.config-file=/etc/prometheus/thanos_objstore.yml --rule-file=/etc/prometheus/alerts/thanos_rule/*.yml --rule-file=/etc/prometheus/rules/thanos_rule/*.yml --alert.query-url=http://thanos-a-1.phl/thanos --alertmanagers.url=http://localhost:9093 --label=prometheus_server="thanos-a-1.phl-rule" --alert.label-drop=prometheus_server --web.external-prefix=/thanos-rule --web.route-prefix=/thanos-rule --query=localhost:19090/thanos
```
*Thanos Store*
```
/usr/bin/docker run --rm --name thanos-store --network host --volume /etc/prometheus/thanos_objstore.yml:/thanos_objstore.yml --volume thanos-data:/data bitnami/thanos:0.23.1 store --data-dir=/data/thanos/store --grpc-address=0.0.0.0:19101 --http-address=0.0.0.0:19201 --objstore.config-file=/thanos_objstore.yml
```
*Thanos Query*
```
/usr/bin/docker run --rm --name thanos-query --network host bitnami/thanos:0.23.1 query --query.replica-label=prometheus_server --grpc-address=0.0.0.0:19102 --http-address=0.0.0.0:19090 --web.external-prefix=/thanos --store=localhost:19100 --store=localhost:19101 --store=localhost:19104
```
*Thanos Compact*
```
/usr/bin/docker run --rm --name thanos-compact --network host --volume /etc/prometheus/thanos_objstore.yml:/thanos_objstore.yml --volume thanos-data:/data bitnami/thanos:0.23.1 compact --data-dir=/data/thanos/compact --http-address=0.0.0.0:19203 --objstore.config-file=/thanos_objstore.yml --retention.resolution-raw=548d --retention.resolution-5m=548d --retention.resolution-1h=548d --wait
```

**Comment by yeya24** [comment]
Is sidecar uploading new blocks to the same bucket that Thanos store is using?

**Comment by lpashos** [comment]
Yes, sidecar (and rule) is uploading new blocks to the same object store that thanos store is looking. All thanos services that require an object store configuration share the same configuration file to ensure that they are all pointed to the same place.

Additionally, I managed to find a workaround that got thanos store accepting new blocks again, but it may not be viable for some setups. I created a new object storage container and pointed our object storage configuration at it. I created a new docker volume for all prometheus and thanos data and copied all prometheus/thanos data over to it while ommitting all "thanos.shipper.json" files so that thanos-sidecar would upload all existing prometheus and thanos-rule blocks to the new swift container. Essentially this created a fresh installation of thanos while retaining all data from the last 15 days. I could see that thanos-sidecar uploaded all available blocks and thanos-store indexed them without issues. I then spun up a second instance of thanos store that was pointed at the old swift container and added it to the thanos-query configuration in order to retain as much of our historical data as possible for queries. This new setup has run well for the past few days.

Based upon this, I suspect that thanos-store is choking on something in our old object storage container, but I cannot see what the problem is from the available logs. Here's the full initialization sequence with the log level in debug mode from the store pointed at our old data. It should be finding many more blocks than it is.
[store.log](https://github.com/thanos-io/thanos/files/7705842/store.log)

**Comment by Lykos153** [comment]
We had a similar issue. Our metrics are stored in a ceph-rgw and accessed by thanos-store via swift. We observed that the block count according to thanos-store was stuck at 808 while we found that new blocks were still uploaded to the bucket (accumulating to ~3000 blocks). Thanks to @lpashos's hint we added some debug output to `fetchMetadata` in `fetcher.go` to find the "choke point", but to no avail. After switching to a S3 based connection to the same ceph-rgw, thanos-store fetched all blocks as expected.

TL;DR: Switching from Swift to S3 solved the issue for us.

`thanos-store v0.26.0`
(We don't know the version of the remote ceph-rgw, unfortunately)

CC: @expektorans
